### PR TITLE
Add clBLAS and clBLAST benchmarks for TRSM

### PR DESCRIPTION
**Pull request by mcleary**
Add TRSM benchmarks for both clBLAS and clBLAST

### CODE REVIEW

File: benchmark/clblas/clwrap.h [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+
+      if (status != CL_SUCCESS) {
+        do_error("failure to release memobject");
+      }
+      is_active = false;
+    }
+  }
+};
+
+#endif /* end of include guard: CLWRAP_HPP */
```
Comment (CodeplayBen) [inline benchmark/clblas/clwrap.h:201]:
```suggestion
#endif /* end of include guard: CLWRAP_HPP */

```

File: benchmark/clblas/main.cpp [hunk 2]
```diff
@@ -0,0 +1,38 @@
+#include "utils.hpp"
```
Comment (CodeplayBen) [inline benchmark/clblas/main.cpp:1]: Does this also need a license header?

File: benchmark/clblas/utils.hpp [hunk 3]
```diff
@@ -0,0 +1,169 @@
+#ifndef CLBLAST_UTILS_HPP
```
Comment (CodeplayBen) [inline benchmark/clblas/utils.hpp:1]: License header?

File: benchmark/clblas/utils.hpp [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+  } else {
+    // Return a really big number to show that we've failed.
+    return 0xFFFFFFFFFFFFFFFF;
+  }
+}
+
+}  // namespace utils
+}  // namespace blas_benchmark
+
+#endif
```
Comment (CodeplayBen) [inline benchmark/clblas/utils.hpp:169]:
```suggestion
#endif

```

File: benchmark/clblast/blas3/trsm.cpp [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+  std::vector<data_t> a(sizeA);
+  std::vector<data_t> b = blas_benchmark::utils::random_data<data_t>(sizeB);
+
+  const scalar_t diagValue =
+      diagonal == 'u' ? data_t{1}
+                      : blas_benchmark::utils::random_scalar<scalar_t>(
+                            scalar_t{1}, scalar_t{10});
+
+  blas_benchmark::utils::fill_trsm_matrix(a, k, lda, triangle, diagValue,
+                                          scalar_t{0});
```
Comment (CodeplayBen) [inline benchmark/clblast/blas3/trsm.cpp:62]: Same as the clblas benchmark these should maybe be ```data_t``` though that might not be all that's needed for half to work with these.

File: benchmark/syclblas/blas3/trsm.cpp [hunk 6]
```diff
@@ -108,9 +111,35 @@ void run(benchmark::State& state, ExecutorType* executorPtr, char side,
 
   // Measure
   for (auto _ : state) {
-    // Run
+    auto start = std::chrono::system_clock::now();
+    auto events = blas_method_def();
+    auto end = std::chrono::system_clock::now();
+    double overall_time = (end - start).count();
```
Comment (CodeplayBen) [inline benchmark/syclblas/blas3/trsm.cpp:117]: Would ```std::chrono::high_resolution_clock``` not be preferable for this?

**Review by CodeplayBen — COMMENTED** [review]
Looks good just some small comments

File: benchmark/clblas/clwrap.cpp [hunk 7]
```diff
@@ -0,0 +1,331 @@
...
+ *  @filename clwrap.cpp
+ *
+ **************************************************************************/
+
+#include "clwrap.h"
+
+#include <algorithm>
+#include <memory>
+
+char *getCLErrorString(cl_int err) {
```
Comment (ProGTX) [inline benchmark/clblas/clwrap.cpp:31]: Where does this file come from? I thought the user would have to provide their own `clblas`?

File: benchmark/clblas/clwrap.cpp [hunk 8]
```diff
@@ -0,0 +1,331 @@
...
+  }
+  return score;
+}
+
+int OpenCLDeviceSelector::score_device(std::string requested,
+                                       cl_device_id device) {
+  // Get the requested device type:
+  cl_device_type req_type = match_device_type(requested);
+  cl_device_type dev_type;
+  cl_int status = clGetDeviceInfo(device, CL_DEVICE_TYPE,
```
[the remark this answers is not in the document]
Comment (mcleary) [reply benchmark/clblas/clwrap.cpp:177]: The `clwrap` code is duplicated in the repo. My suggestion is to make this into a library to be used by all benchmarks that need it, in a separate pull request.

File: benchmark/clblas/clwrap.cpp (hunk 7 above)
Comment (mcleary) [reply benchmark/clblas/clwrap.cpp:31]: I'm not sure if I understood the question

File: benchmark/clblas/blas3/trsm.cpp [hunk 9]
```diff
@@ -0,0 +1,216 @@
...
+  const int sizeA = k * lda;
+  const int sizeB = n * ldb;
+
+  // Matrices
+  std::vector<data_t> a(sizeA);
+  std::vector<data_t> b = blas_benchmark::utils::random_data<data_t>(sizeB);
+
+  const scalar_t diagValue =
+      diagonal == 'u' ? data_t{1}
+                      : blas_benchmark::utils::random_scalar<scalar_t>(
```
[the remark this answers is not in the document]
Comment (mcleary) [reply benchmark/clblas/blas3/trsm.cpp:34]: I changed the type do `data_t`

File: benchmark/clblas/blas3/trsm.cpp [hunk 10]
```diff
@@ -0,0 +1,216 @@
+
```
[the remark this answers is not in the document]
Comment (mcleary) [reply benchmark/clblas/blas3/trsm.cpp:1]: Done

File: benchmark/clblas/main.cpp (hunk 2 above)
Comment (mcleary) [reply benchmark/clblas/main.cpp:1]: Done

File: benchmark/clblas/utils.hpp (hunk 3 above)
Comment (mcleary) [reply benchmark/clblas/utils.hpp:1]: Done

File: benchmark/clblast/blas3/trsm.cpp (hunk 5 above)
Comment (mcleary) [reply benchmark/clblast/blas3/trsm.cpp:62]: Done

File: benchmark/syclblas/blas3/trsm.cpp (hunk 6 above)
Comment (mcleary) [reply benchmark/syclblas/blas3/trsm.cpp:116]: Yes, however, the `system_clock` in https://github.com/codeplaysoftware/sycl-blas/blob/2e0e5e90e27b393c9adf47c5ec26a8376158e4fc/benchmark/common_utils.hpp#L625) so I kept the same here.

We could add this to the internal discussion on sycl-blas

Comment (CodeplayBen) [reply benchmark/syclblas/blas3/trsm.cpp:116]: That's fair, I think it's fine to leave it as is then.

File: benchmark/clblas/clwrap.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (CodeplayBen) [reply benchmark/clblas/clwrap.cpp:177]: I'm not sure I understand. I meant that all other calls to OpenCL functions check the returned status for errors but this one doesn't, is that intentional?

[the remark this answers is not in the document]
Comment (mcleary) [reply benchmark/clblas/clwrap.cpp:177]: I meant that the clwrap.cpp is duplicated in the clblas and clblast folder. So I think I will turn this into a library that can be used by the benchmarks that need it.

but you are right, it should be checked

[the remark this answers is not in the document]
Comment (mcleary) [reply benchmark/clblas/clwrap.cpp:177]: See https://github.com/codeplaysoftware/sycl-blas/pull/274 for a fix of this part

File: benchmark/clblas/clwrap.cpp (hunk 7 above)
Comment (mcleary) [reply benchmark/clblas/clwrap.cpp:31]: See https://github.com/codeplaysoftware/sycl-blas/pull/274 for a proposal to make a common library with this code

Comment (ProGTX) [reply benchmark/clblas/clwrap.cpp:31]: Oh, I see, I thought this file came from `clblas`, that PR makes a lot of sense, thanks

**Review by ProGTX — APPROVED** [review]

**Review by Bensuo — APPROVED** [review]

**Pull request merged**